### PR TITLE
#47: Added parquet support

### PR DESCRIPTION
**Pull request by jakobbraun**

### CODE REVIEW

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/LoadedFile.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+     * </p>
+     * 
+     * @return {@link RandomAccessInputStream} of the file.
+     */
+    public RandomAccessInputStream getRandomAccessInputStream() {
+        try (final InputStream inputStream = getInputStream()) {
+            return InMemoryRandomAccessStream.getInMemoryCache(inputStream);
+        } catch (final IOException exception) {
+            throw new IllegalStateException(
+                    ExaError.messageBuilder("E-VSDF-11").message("Failed an data sources input stream.").toString(),
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/documentfetcher/files/LoadedFile.java:46]:
```suggestion
                    ExaError.messageBuilder("E-VSDF-11").message("Failed reading data source's input stream.").toString(),
```

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/parquet/ParquetDocumentFetcher.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    protected ParquetDocumentFetcher(final StringFilter filePattern, final SegmentDescription segmentDescription,
+            final FileLoaderFactory fileLoaderFactory) {
+        super(filePattern, segmentDescription, fileLoaderFactory);
+    }
+
+    @Override
+    protected Stream<DocumentNode<AvroNodeVisitor>> readDocuments(final LoadedFile loadedFile) {
+        final InputFile hadoopInputFile = SeekableInputStreamAdapter.convert(loadedFile.getRandomAccessInputStream());
+        try {
+            final ParquetReader<GenericRecord> reader = AvroParquetReader.<GenericRecord>builder(hadoopInputFile)
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/documentfetcher/files/parquet/ParquetDocumentFetcher.java:41]: Note to self: InputFile should also be supported in the parquet reader library also.

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/randomaccessinputstream/InMemoryRandomAccessStream.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+     *
+     * @param normalInputStream input stream to read from
+     * @return built {@link InMemoryRandomAccessStream}
+     */
+    public static InMemoryRandomAccessStream getInMemoryCache(final InputStream normalInputStream) {
+        try {
+            return new InMemoryRandomAccessStream(normalInputStream.readAllBytes());
+        } catch (final IOException exception) {
+            throw new IllegalStateException(ExaError.messageBuilder("E-VSDF-12").message(
+                    "This data source does not support random access for files. So we tried to keep everything in memory, but it failed (probably due to too few memory).")
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/documentfetcher/files/randomaccessinputstream/InMemoryRandomAccessStream.java:31]: We should write this to user guide that it is good practice to have < 1GB parquet files. UDFs also have limited memory.

File: src/test/java/com/exasol/adapter/document/documentfetcher/files/CloseCheckLoadedFile.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package com.exasol.adapter.document.documentfetcher.files;
+
+import java.io.InputStream;
+
+class CloseCheckLoadedFile extends LoadedFile {
```
Comment (morazow) [inline src/test/java/com/exasol/adapter/document/documentfetcher/files/CloseCheckLoadedFile.java:5]: Suggestion:
`CloseableLoadedFile`
`CloseableStream`

File: src/test/java/com/exasol/adapter/document/documentfetcher/files/CloseCheckLoadedFile.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        super("");
+        this.closeCheckStream = new CloseCheckStream(content);
+    }
+
+    @Override
+    public InputStream getInputStream() {
+        return this.closeCheckStream;
+    }
+
+    public boolean wasStreamClosed() {
```
Comment (morazow) [inline src/test/java/com/exasol/adapter/document/documentfetcher/files/CloseCheckLoadedFile.java:18]:
```suggestion
    public boolean isStreamClosed() {
```

**Review by morazow — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/randomaccessinputstream/InMemoryRandomAccessStream.java (hunk 3 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/document/documentfetcher/files/randomaccessinputstream/InMemoryRandomAccessStream.java:31]: Yes, but only for data sources that don't support random access read. This class is only used as a polyfill for those data sources (Currently none; only for testing).

**Review by morazow — APPROVED** [review]

**Pull request merged**